### PR TITLE
Split Lovelace out to new module and add multi-asset Value types

### PR DESCRIPTION
**Pull request by dcoutts**
Next step is to use the multi-asset Value.

This takes a slightly different tack on how to represent multi-asset values, namely by making 'AssetId' a disjoint sum so that ada has a label distinct from all other custom assets.
```
data AssetId = AssetId !PolicyId !AssetName
             | AdaAssetId

data Value

selectAsset :: Value -> (AssetId -> Quantity)
valueFromList :: [(AssetId, Quantity)] -> Value
valueToList :: Value -> [(AssetId, Quantity)]
```
plus derived convenience functions
```
selectLovelace :: Value -> Lovelace
lovelaceToValue :: Lovelace -> Value
```

### CODE REVIEW

File: cardano-api/src/Cardano/Api/Value.hs [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+     TxOutValue    :: MultiAssetInEra era -> Value -> TxOutValue era
+
+deriving instance Eq   (TxOutValue era)
+deriving instance Show (TxOutValue era)
+
+
+-- | Representation of whether multi-asset transactions are supported in a
+-- particular era.
+--
+data AdaOnlyInEra era where
```
Comment (intricate) [inline cardano-api/src/Cardano/Api/Value.hs:175]: Is this the intended haddock for this type? or a copy-paste mistake from the `MultiAssetInEra` type?

File: cardano-api/src/Cardano/Api/Value.hs [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+
+deriving instance Eq   (MintValue era)
+deriving instance Show (MintValue era)
+
+
+data TxOutValue era where
+
+     TxOutLovelace :: AdaOnlyInEra era -> Lovelace -> TxOutValue era
+
+     TxOutValue    :: MultiAssetInEra era -> Value -> TxOutValue era
```
Comment (intricate) [inline cardano-api/src/Cardano/Api/Value.hs:166]: Just making sure: we're not intending to support `TxOutLovelace` tx outs for Mary? So clients will need to use the `Value` type for Mary transactions regardless?

File: cardano-api/src/Cardano/Api/Value.hs (hunk 1 above)
Comment (dcoutts) [reply cardano-api/src/Cardano/Api/Value.hs:175]: Oops. Yup.

File: cardano-api/src/Cardano/Api/Value.hs (hunk 2 above)
Comment (dcoutts) [reply cardano-api/src/Cardano/Api/Value.hs:166]: Yes. And one can make a `Value` from just `Lovelace`.

Comment (intricate) [reply cardano-api/src/Cardano/Api/Value.hs:166]: Ah okay, I see you've got a function, `lovelaceToValue :: Lovelace -> Value`. That should make it easy if we're going to allow users to specify simple lovelace-only tx outs for Mary transactions in the CLI or some other client application.

**Review by intricate — APPROVED** [review]

**Review by newhoggy — APPROVED** [review]

File: cardano-api/src/Cardano/Api/Value.hs [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+  deriving (Eq, Show)
+
+instance Semigroup Value where
+  Value a <> Value b = Value (mergeAssetMaps a b)
+
+instance Monoid Value where
+  mempty = Value Map.empty
+
+
+{-# NOINLINE mergeAssetMaps #-}
```
Comment (newhoggy) [inline cardano-api/src/Cardano/Api/Value.hs:106]: What's the reason for `NOINLINE`?

File: cardano-api/src/Cardano/Api/Value.hs [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+valueToList :: Value -> [(AssetId, Quantity)]
+valueToList (Value m) = Map.toList m
+
+-- | This lets you write @a - b@ as @a <> negateValue b@.
+--
+negateValue :: Value -> Value
+negateValue (Value m) = Value (Map.map negate m)
+
+filterValue :: (AssetId -> Bool) -> Value -> Value
+filterValue p (Value m) = Value (Map.filterWithKey (\k _v -> p k) m)
```
Comment (newhoggy) [inline cardano-api/src/Cardano/Api/Value.hs:139]: I would probably call this `filterValueByAssetId`

File: cardano-api/src/Cardano/Api/Value.hs (hunk 3 above)
Comment (dcoutts) [reply cardano-api/src/Cardano/Api/Value.hs:106]: Thanks. Noted in a comment.

**Comment by dcoutts** [comment]
bors merge

File: cardano-api/src/Cardano/Api/Value.hs (hunk 3 above)
Comment (newhoggy) [reply cardano-api/src/Cardano/Api/Value.hs]: Thanks!

**Pull request merged**